### ISSUE TITLE
Build fails with BUILD_ISPC_MODULE ON

### ISSUE THREAD

**Issue by xkaraman** [isl-org/Open3D#4058]
**Describe the bug**
Trying to build from source.  It fails with
Error: Unknown option "-pthread". 
cpp/open3d/utility/CMakeFiles/utility.dir/build.make:222: recipe for target 'cpp/open3d/utility/CMakeFiles/utility.dir/ISAInfo.ispc.o' failed
make[2]: *** [cpp/open3d/utility/CMakeFiles/utility.dir/ISAInfo.ispc.o] Error 255
CMakeFiles/Makefile2:3613: recipe for target 'cpp/open3d/utility/CMakeFiles/utility.dir/all' failed
make[1]: *** [cpp/open3d/utility/CMakeFiles/utility.dir/all] Error 2

**To Reproduce**
activate conda evnironment with dependencies installed
mkdir build && cd build
cmake .. (CUDA_MODULE OFF)
make -j8

**Expected behavior**
A successfull build.

**Screenshots**

**Environment (please complete the following information):**

- Operating system: Ubuntu 18.04
- Python version: Python 3.9.6
- Open3D version: 0.13
- Is this remote workstation?: no
- How did you install Open3D?: build from source
- Compiler version (if built from source):  gcc 7.5,

**Additional context**
If i set BUILD_ISPC_MODULE to OFF then build ends succesfully.

**Comment by stotko** [comment]
Thanks for the report. We are aware of this limitation and handled this case when we introduced ISPC support here: https://github.com/isl-org/Open3D/blob/0a405eac7386afffea42368112ffadf1c94b5b50/CMakeLists.txt#L314

Could you provide the complete console output when running CMake, i.e. when running `cmake ..`?

**Comment by xkaraman** [comment]
`cmake .. -DCMAKE_BUILD_TYPE=Release`

Console output:

-- CMAKE_BUILD_TYPE is set to Release.
-- Downloading 3rdparty dependencies to /home/xenofon/git/Open3D/3rdparty_downloads
CMake Deprecation Warning at CMakeLists.txt:137 (cmake_policy):
  The OLD behavior for policy CMP0072 will be removed from a future version
  of CMake.

  The cmake-policies(7) manual explains that the OLD behaviors of all
  policies are deprecated and that a policy should be set to OLD only under
  specific short-term circumstances.  Projects should be ported to the NEW
  behavior and not rely on setting a policy to OLD.


-- Open3D 0.13.0+0a405eac7
-- Using security hardening compiler flags: $<$<COMPILE_LANGUAGE:CUDA>:-Xcompiler=-fstack-protector-strong,-Wformat,-Wformat-security>;$<$<COMPILE_LANGUAGE:CXX>:-fstack-protector-strong;-Wformat;-Wformat-security>
-- Using security hardening linker flags: $<$<COMPILE_LANGUAGE:CUDA>:-Xcompiler=-Wl,-z,noexecstack,-Wl,-z,relro,-z,now,$<$<STREQUAL:$<TARGET_PROPERTY:TYPE>,EXECUTABLE>:-pie>,$<$<CONFIG:Release>:LINKER:--strip-debug>>;$<$<COMPILE_LANGUAGE:CXX>:-Wl,-z,noexecstack;-Wl,-z,relro,-z,now;$<$<STREQUAL:$<TARGET_PROPERTY:TYPE>,EXECUTABLE>:-pie>;$<$<CONFIG:Release>:LINKER:--strip-debug>>
-- Using security hardening compiler definitions: _FORTIFY_SOURCE=2
-- Found CUDAToolkit: /usr/local/cuda/include  
-- Building with detected architectures
-- Using CUDA architectures: 61-real
-- Using user-provided path to the ISPC compiler
-- Using ISPC instruction sets: sse2-i32x4;sse4-i32x4;avx1-i32x8;avx2-i32x8;avx512knl-i32x16;avx512skx-i32x16
-- Found CUDAToolkit: /usr/local/cuda/include (found version "11.4.120") 
-- Using installed third-party library 3rdparty_threads 
-- Using installed third-party library 3rdparty_openmp 
-- Building with OpenMP
-- Using installed third-party library 3rdparty_x11 
-- Building library 3rdparty_glfw from source
-- Using X11 for window creation
-- Building third-party library JPEG from source
-- NASM assembler not found - libjpeg-turbo performance may suffer
-- libturbojpeg: WITH_CRT_DLL=OFF
-- pybind11 v2.6.2 
-- Failed to find LLVM FileCheck
-- git version: v0.12.0-449-g0a405eac normalized to 0.12.0.449
-- Version: 0.12.0.449
-- Performing Test HAVE_STD_REGEX -- success
-- Performing Test HAVE_GNU_POSIX_REGEX -- failed to compile
-- Performing Test HAVE_POSIX_REGEX -- success
-- Performing Test HAVE_STEADY_CLOCK -- success
Perf Counters support requested, but was unable to find libpfm.
-- Using installed third-party library 3rdparty_opengl 
-- Using prebuilt third-party library Filament
-- GLIBC version 2.27-3 found: Using Google Filament binary.
-- Filament is located at /home/xenofon/git/Open3D/build/filament/src/ext_filament
-- CLANG_LIBDIR: /usr/lib/x86_64-linux-gnu
-- CPP_LIBRARY: /usr/lib/x86_64-linux-gnu/libc++.so
-- CPPABI_LIBRARY: /usr/lib/x86_64-linux-gnu/libc++abi.so
-- Building third-party library faiss from source
-- Using MKL to support BLAS and LAPACK functionalities.
-- Using installed third-party library 3rdparty_cuda_npp 11.4.120
-- IPP-ICV 2020.0.0 Gold available. Building interface wrappers IPP-IW.
-- Copying /home/xenofon/git/Open3D/cpp/open3d/visualization/webrtc_server/html to /home/xenofon/git/Open3D/build/bin/resources.
-- PYPI_PACKAGE_NAME: open3d
-- ================================================================================
-- Open3D 0.13.0 Configuration Summary
-- ================================================================================
-- Enabled Features:
--   OpenMP .................................. ON
--   Headless Rendering ...................... OFF
--   Azure Kinect Support .................... OFF
--   Intel RealSense Support ................. ON
--   CUDA Support ............................ ON
--   ISPC Support ............................ ON
--   Build GUI ............................... ON
--   Build WebRTC visualizer ................. ON
--   Build Shared Library .................... OFF
--   Build Unit Tests ........................ ON
--   Build Examples .......................... ON
--   Build Python Module ..................... ON
--   Build Jupyter Extension ................. ON
--   Build TensorFlow Ops .................... OFF
--   Build PyTorch Ops ....................... OFF
--   Build Benchmarks ........................ ON
--   Bundle Open3D-ML ........................ OFF
--   Force GLIBCXX_USE_CXX11_ABI= ............ 0
-- ================================================================================
-- Third-Party Dependencies:
--   Eigen3 .................................. yes (build from source)
--   faiss ................................... yes (build from source)
--   filament ................................ yes (build from source)
--   fmt ..................................... yes (build from source)
--   GLEW .................................... yes (build from source)
--   GLFW .................................... yes (build from source)
--   imgui ................................... yes (build from source)
--   ippicv .................................. yes (build from source)
--   JPEG .................................... yes (build from source)
--   jsoncpp ................................. yes (build from source)
--   liblzf .................................. yes (build from source)
--   OpenGL .................................. yes (build from source)
--   PNG ..................................... yes (build from source)
--   qhullcpp ................................ yes (build from source)
--   librealsense ............................ yes (build from source)
--   tinyfiledialogs ......................... yes (build from source)
--   TinyGLTF ................................ yes (build from source)
--   tinyobjloader ........................... yes (build from source)
--   WebRTC .................................. yes (build from source)

By handled you mean it should not error out?

**Comment by stotko** [comment]
Yes, the error should not occur as setting that flag to `FALSE` means "do not use `-pthread` flag". So, the error you are observing should actually not occur.

Is this log from a clean build, i.e. with `build` directory removed before running CMake? Some lines that come from a clean build are missing, e.g. compiler detection. Could you check this and provide the log from a clean build?

**Comment by xkaraman** [comment]
It was after make clean. But i will try it again and respond. 

Another problem i encountered is that even after a successful build, when trying to `import open3d` the following is thrown
`Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/xenofon/miniconda3/envs/open3dbuild/lib/python3.9/site-packages/open3d/__init__.py", line 69, in <module>
    from open3d.cuda.pybind import (camera, geometry, io, pipelines,
ModuleNotFoundError: No module named 'open3d.cuda.pybind'
`

**Comment by xkaraman** [comment]
`cmake ..` output after removing build dir and recreating it

-- CMAKE_BUILD_TYPE is set to Release.
-- Downloading 3rdparty dependencies to /home/xenofon/git/Open3D/3rdparty_downloads
CMake Deprecation Warning at CMakeLists.txt:137 (cmake_policy):
  The OLD behavior for policy CMP0072 will be removed from a future version
  of CMake.

  The cmake-policies(7) manual explains that the OLD behaviors of all
  policies are deprecated and that a policy should be set to OLD only under
  specific short-term circumstances.  Projects should be ported to the NEW
  behavior and not rely on setting a policy to OLD.


-- The C compiler identification is GNU 7.5.0
-- The CXX compiler identification is GNU 7.5.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Open3D 0.13.0+0a405eac7
-- Found Python3: /home/xenofon/miniconda3/envs/open3dbuild/bin/python3.9 (found suitable version "3.9.6", required range is "3.6...<3.10") found components: Interpreter Development Development.Module Development.Embed 
-- Performing Test FLAG_fstack_protector_strong
-- Performing Test FLAG_fstack_protector_strong - Success
-- Performing Test FLAG_Wformat
-- Performing Test FLAG_Wformat - Success
-- Performing Test FLAG_Wformat_security
-- Performing Test FLAG_Wformat_security - Success
-- Performing Test FLAG_Wl__z_noexecstack
-- Performing Test FLAG_Wl__z_noexecstack - Success
-- Performing Test FLAG_Wl__z_relro__z_now
-- Performing Test FLAG_Wl__z_relro__z_now - Success
-- Performing Test FLAG_pie
-- Performing Test FLAG_pie - Success
-- Performing Test FLAG____CONFIG_Release__LINKER___strip_debug_
-- Performing Test FLAG____CONFIG_Release__LINKER___strip_debug_ - Success
-- Using security hardening compiler flags: $<$<COMPILE_LANGUAGE:CUDA>:-Xcompiler=-fstack-protector-strong,-Wformat,-Wformat-security>;$<$<COMPILE_LANGUAGE:CXX>:-fstack-protector-strong;-Wformat;-Wformat-security>
-- Using security hardening linker flags: $<$<COMPILE_LANGUAGE:CUDA>:-Xcompiler=-Wl,-z,noexecstack,-Wl,-z,relro,-z,now,$<$<STREQUAL:$<TARGET_PROPERTY:TYPE>,EXECUTABLE>:-pie>,$<$<CONFIG:Release>:LINKER:--strip-debug>>;$<$<COMPILE_LANGUAGE:CXX>:-Wl,-z,noexecstack;-Wl,-z,relro,-z,now;$<$<STREQUAL:$<TARGET_PROPERTY:TYPE>,EXECUTABLE>:-pie>;$<$<CONFIG:Release>:LINKER:--strip-debug>>
-- Using security hardening compiler definitions: _FORTIFY_SOURCE=2
-- Fetching ISPC compiler
-- Using ISPC instruction sets: sse2-i32x4;sse4-i32x4;avx1-i32x8;avx2-i32x8;avx512knl-i32x16;avx512skx-i32x16
-- The ISPC compiler identification is Intel 1.16.1
-- Detecting ISPC compiler ABI info
-- Detecting ISPC compiler ABI info - done
-- Check for working ISPC compiler: /home/xenofon/git/Open3D/build/_deps/ext_ispc-src/bin/ispc - skipped
-- Performing Test HAS_GLIBCXX_USE_CXX11_ABI
-- Performing Test HAS_GLIBCXX_USE_CXX11_ABI - Success
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Using installed third-party library 3rdparty_threads 
-- Found OpenMP_C: -fopenmp (found version "4.5") 
-- Found OpenMP_CXX: -fopenmp (found version "4.5") 
-- Found OpenMP: TRUE (found version "4.5")  
-- Using installed third-party library 3rdparty_openmp 
-- Building with OpenMP
-- Using installed third-party library 3rdparty_x11 
-- Building library 3rdparty_glfw from source
-- Found Vulkan: /usr/lib/x86_64-linux-gnu/libvulkan.so  
-- Using X11 for window creation
-- Found X11: /usr/include   
-- Building third-party library JPEG from source
-- Looking for a ASM_NASM compiler
-- Looking for a ASM_NASM compiler - NOTFOUND
-- NASM assembler not found - libjpeg-turbo performance may suffer
-- libturbojpeg: WITH_CRT_DLL=OFF
-- pybind11 v2.6.2 
-- Performing Test HAS_FLTO
-- Performing Test HAS_FLTO - Success
-- Found OpenGL: /usr/lib/x86_64-linux-gnu/libOpenGL.so   
-- Using installed third-party library 3rdparty_opengl 
-- Using prebuilt third-party library Filament
-- GLIBC version 2.27-3 found: Using Google Filament binary.
-- Filament is located at /home/xenofon/git/Open3D/build/filament/src/ext_filament
-- CLANG_LIBDIR: /usr/lib/x86_64-linux-gnu
-- CPP_LIBRARY: /usr/lib/x86_64-linux-gnu/libc++.so
-- CPPABI_LIBRARY: /usr/lib/x86_64-linux-gnu/libc++abi.so
-- Building third-party library faiss from source
-- Using MKL to support BLAS and LAPACK functionalities.
-- IPP-ICV 2020.0.0 Gold available. Building interface wrappers IPP-IW.
-- Copying /home/xenofon/git/Open3D/cpp/open3d/visualization/webrtc_server/html to /home/xenofon/git/Open3D/build/bin/resources.
-- PYPI_PACKAGE_NAME: open3d
-- ================================================================================
-- Open3D 0.13.0 Configuration Summary
-- ================================================================================
-- Enabled Features:
--   OpenMP .................................. ON
--   Headless Rendering ...................... OFF
--   Azure Kinect Support .................... OFF
--   Intel RealSense Support ................. OFF
--   CUDA Support ............................ OFF
--   ISPC Support ............................ ON
--   Build GUI ............................... ON
--   Build WebRTC visualizer ................. ON
--   Build Shared Library .................... OFF
--   Build Unit Tests ........................ OFF
--   Build Examples .......................... ON
--   Build Python Module ..................... ON
--   Build Jupyter Extension ................. OFF
--   Build TensorFlow Ops .................... OFF
--   Build PyTorch Ops ....................... OFF
--   Build Benchmarks ........................ OFF
--   Bundle Open3D-ML ........................ OFF
--   Force GLIBCXX_USE_CXX11_ABI= ............ 0
-- ================================================================================
-- Third-Party Dependencies:
--   Eigen3 .................................. yes (build from source)
--   faiss ................................... yes (build from source)
--   filament ................................ yes (build from source)
--   fmt ..................................... yes (build from source)
--   GLEW .................................... yes (build from source)
--   GLFW .................................... yes (build from source)
--   imgui ................................... yes (build from source)
--   ippicv .................................. yes (build from source)
--   JPEG .................................... yes (build from source)
--   jsoncpp ................................. yes (build from source)
--   liblzf .................................. yes (build from source)
--   OpenGL .................................. yes (build from source)
--   PNG ..................................... yes (build from source)
--   qhullcpp ................................ yes (build from source)
--   librealsense ............................ no
--   tinyfiledialogs ......................... yes (build from source)
--   TinyGLTF ................................ yes (build from source)
--   tinyobjloader ........................... yes (build from source)
--   WebRTC .................................. yes (build from source)
-- ================================================================================

**Comment by stotko** [comment]
Thanks. Does the clean build succeed or does it still produce the `-pthread` flag error? The log looks good and the parts related to the `-pthread` flag are the same as on my machine (Ubuntu 20.04).

The pybind issue is unrelated to the ISPC build error. Is this from an older build of Open3D, e.g. the release or dev wheel, or from your current build, e.g. via `pip uninstall open3d --yes && make install-pip-package`?

**Comment by xkaraman** [comment]
About the pybind. It's the latest from master branch  after successfully building without ISPC and with CUDA module on. i did both instructions you said `pip uninstall open3d && make install-pip-package` and when trying to import it the error is thrown.

Build seems to succeed with no errors. Does it matter if i `make` instead of `make -j8`? But currently other modules i wanted are off (like benchmark,cuda,realsense,jupyter).

Should i rebuild them after a clean build dir?

**Comment by stotko** [comment]
Could you open another issue for the pybind problem? It seems unrelated to ISPC support, so it is better to discuss this separately.

Good to hear that the build now succeeds. `make` and `make -j8` only differ in that the former builds with a single process whereas the latter uses up to 8 processes in parallel to speed-up compilation. Despite that parallelization, they are equivalent.

If you want to build with other options and modules enabled/disabled, just add the options when running cmake, e.g. `cmake -DBUILD_CUDA_MODULE=ON ..`. It is highly recommended to always do a clean build when changing any such configuration options to ensure a consistent build.

**Comment by xkaraman** [comment]
Yeah i know the parallelization but that was the difference apart from the clean build, so maybe it was something to note.
Anyway, import now works, also in jupyter-lab. 
Last question before closing the issue? 
What is the difference on `open3d.core.cuda` and `open3d.cuda`

**Comment by stotko** [comment]
Depending on whether CUDA support is enabled, the pybind symbols are either available under `open3d.cuda.pybind` or `open3d.cpu.pybind`. This is an implementation detail, so all symbols get automatically imported into the `open3d` namespace, e.g. `open3d.cuda.pybind.core.cuda` -> `open3d.core.cuda`.

**Comment by xkaraman** [comment]
Oh alrighty. Thanks for the info! Closing the issue since the original issue is fixed!
